File: Assets/Import_out/Uzil/API/Audio.cs
using System;
using System.Collections.Generic;

using Uzil;
using Uzil.Audio;

namespace UZAPI {

public class Audio {

	/*======================================Constructor==========================================*/

	/*=====================================Static Members========================================*/

	/*=====================================Static Funciton=======================================*/

	/*=========================================Members===========================================*/

	/*========================================Components=========================================*/

	/*==========================================Event============================================*/

	/*======================================Unity Function=======================================*/

	/*========================================Interface==========================================*/

	/** 設置音量 */
	public static void SetVolumeLinear (string mixerName, float volume) {
		AudioUtil.SetVolumeLinear(mixerName, volume);
	}

	/** 取得音量 */
	public static float GetVolumeLinear (string mixerName) {
		float aa = AudioUtil.GetVolumeLinear(mixerName);
		return aa;
	}

	/** 是否存在 */
	public static bool IsExist (string id) {
		return AudioMgr.Inst().IsExist(id);
	}

	/** 預載 */
	public static string Preload (string id, string pathOrKey, string argsJson) {
		DictSO args = DictSO.Json(argsJson);
		if (args == null) args = new DictSO();

		if (pathOrKey == null) return null;

		AudioObj audio = AudioMgr.Inst().Preload(id, pathOrKey, args);
		if (audio == null) audio = AudioMgr.Inst().Preload(id, AudioSet.Get(pathOrKey), args);
		if (audio == null) return null;

		return audio.id;
	}
	public static void PreloadAsync (string id, string pathOrKey, string argsJson, int callbackID) {
		DictSO args = DictSO.Json(argsJson);
		if (args == null) args = new DictSO();

		Action<string> cb = (audioID)=>{
			Callback.CallLua_cs_arg(callbackID, audioID);
		};

		Action<AudioObj, Action> check = (audio, next)=>{
			if (audio != null) {
				cb(audio.id);
				return;
			}
			if (next != null) next();
		};

		if (pathOrKey == null) {
			check(null, null);
			return;
		}

		AudioMgr audioMgr = AudioMgr.Inst();
		Async.Waterfall(
			new List<Action<Action<bool>>>(){
				(next)=>{
					audioMgr.PreloadAsync(id, pathOrKey, args, (audioObj)=>{
						check(audioObj, ()=>{
							next(true);
						});
					});
				},
				(next)=>{
					audioMgr.PreloadAsync(id, AudioSet.Get(pathOrKey), args, (audioObj)=>{
						check(audioObj, ()=>{
							next(true);
						});
					});
				},
			},
			()=>{
				check(null, ()=>{
					cb(null);
				});
			}
		);
	}

	/** 申請 */
	public static string Request (string id, string pathOrKey, string argsJson) {
		DictSO args = DictSO.Json(argsJson);
		if (args == null) args = new DictSO();

		AudioObj audio = AudioMgr.Inst().Request(id, pathOrKey, args);
		if (audio == null) audio = AudioMgr.Inst().Request(id, AudioSet.Get(pathOrKey), args);
		if (audio == null) return null;

		return audio.id;
	}
	
	/** 設置 */
	public static void Set (string id, string argsJson) {
		DictSO args = DictSO.Json(argsJson);
		if (args == null) args = new DictSO();

		AudioMgr.Inst().Set(id, args);
	}

	/** 播放 */
	public static void Play (string id) {
		AudioMgr.Inst().Play(id);
	}

	/** 播放音效(方便使用) */
	public static string PlaySFX (string pathOrKey, string argsJson = null) {
		DictSO args = DictSO.Json(argsJson);
		if (args == null) args = new DictSO();

		args.Set("mixer", "SFX").Set("isLoop", false).Set("isReleaseOnEnd", true);

		AudioObj audio = AudioMgr.Inst().Emit(pathOrKey, args);
		if (audio == null) audio = AudioMgr.Inst().Emit(AudioSet.Get(pathOrKey), args);
		if (audio == null) return null;

		return audio.id;
	}

	/** 釋放 */
	public static void Release (string id) {
		AudioMgr.Inst().Release(id);
	}

	/** 停止 */
	public static void Stop (string id) {
		AudioMgr.Inst().Stop(id);
	}

	/** 停止所有 */
	public static void StopAll () {
		AudioMgr.Inst().StopAll();
	}

	/** 復原 */
	public static void Resume (string id) {
		AudioMgr.Inst().Resume(id);
	}
	/** 復原所有 */
	public static void ResumeAll () {
		AudioMgr.Inst().ResumeAll();
	}

	/** 暫停 */
	public static void Pause (string id) {
		AudioMgr.Inst().Pause(id);
	}
	/** 暫停所有 */
	public static void PauseAll () {
		AudioMgr.Inst().PauseAll();
	}

	/*=====================================Public Function=======================================*/

	/*===================================Protected Function======================================*/

	/*====================================Private Function=======================================*/

}

}

File: AppData/Mods/BUILT-IN/_lua/LuaScript/Uzil/Audio.lua
local this = {};


--[[ 設置音量 ]]
function this.setVolumeLinear (mixerName, volume)
	UZAPI.Audio.SetVolumeLinear(mixerName, volume);
end

--[[ 取得音量 ]]
function this.getVolumeLinear (mixerName)
	return UZAPI.Audio.GetVolumeLinear(mixerName);
end

--[[ 是否存在 ]] -- bool
function this.isExist(id)
	return UZAPI.Audio.IsExist(id);
end

-- [[ 預載 ]] -- string
function this.preload (preferID, pathOrKey, args)
	return UZAPI.Audio.Preload(preferID, pathOrKey, Json.encode(args));
end

-- [[ 預載 ]] -- void
function this.preloadAsync (preferID, pathOrKey, args, onDone)
	local cbID;
	cbID = Callback.newID(nil, function (id)
		if (onDone ~= nil) then onDone(id) end
	end);
	return UZAPI.Audio.PreloadAsync(preferID, pathOrKey, Json.encode(args), cbID);
end

-- [[ 申請 ]] -- string
function this.request (preferID, pathOrKey, args)
	return UZAPI.Audio.Request(preferID, pathOrKey, Json.encode(args));
end

-- [[ 設置 ]] -- void
function this.set (id, args)
	UZAPI.Audio.Set(id, Json.encode(args));
end

-- [[ 播放 ]] -- void
function this.play (id)
	UZAPI.Audio.Play(id);
end

-- [[ 播放音效(方便使用) ]] -- string
function this.playSFX (pathOrKey, args)
	local args_str = Json.encode(args);
	local id = UZAPI.Audio.PlaySFX(pathOrKey, args_str);
	return id;
end

-- [[ 釋放 ]] -- void
function this.release (id)
	UZAPI.Audio.Release(id);
end

-- [[ 停止 ]] -- void
function this.stop (id)
	UZAPI.Audio.Stop(id);
end
-- [[ 停止所有 ]] -- void
function this.stopAll ()
	UZAPI.Audio.StopAll();
end

-- [[ 復原 ]] -- void
function this.resume (id)
	UZAPI.Audio.Resume(id);
end
-- [[ 復原所有 ]] -- void
function this.resumeAll ()
	UZAPI.Audio.ResumeAll();
end

-- [[ 暫停 ]] -- void
function this.pause (id)
	UZAPI.Audio.Pause(id);
end
-- [[ 暫停所有 ]] -- void
function this.pauseAll ()
	UZAPI.Audio.PauseAll();
end

return this;